### PR TITLE
Error reporting: Display reporter on application segfaults

### LINKED ISSUE #22293: Refactor error_report dialog window to be more self contained.

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
**Description of work**

This PR updates the ErrorReporter dialog so that it catches hard crashes to desktop as well as errors which reach the last chance error handler.

**To test:**

* These tests need to be done on a package version of mantid rather than a local build.
* It also needs to be tested on all OS's

* Open mantid
* Turn usage reporting on/restart mantid
* Run the segfault algorithm and check that the error report dialog appears.
* Check that no report is sent if you select no report
* Check that nothing appears if usage reporting is turned off.

Fixes #22293  . 

**Release Notes** 
Has been added to release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: buildconfig/CMake/DarwinSetup.cmake [hunk 1]
```diff
@@ -100,6 +99,14 @@ set ( CMAKE_INSTALL_PREFIX "" )
 set ( CPACK_PACKAGE_EXECUTABLES MantidPlot )
 set ( INBUNDLE MantidPlot.app/ )
 
+# Copy the launcher script to the correct location
+configure_file ( ${CMAKE_MODULE_PATH}/Packaging/osx/Mantid_osx_launcher 
+                 ${CMAKE_BINARY_DIR}/bin/MantidPlot.app/Contents/MacOS/Mantid_osx_launcher COPYONLY )
+# Copy the error dialog script to the correct location
+configure_file ( ${CMAKE_MODULE_PATH}/Packaging/osx/error_dialog_app.py.in
```
Comment (Mantid-Matthew) [inline buildconfig/CMake/DarwinSetup.cmake]: @keeeto do we still need error_dialog_app.py.in ?

File: scripts/ErrorReporter/error_dialog_app.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+import mantid # noqa
+from ErrorReporter import resources # noqa
+from mantid.kernel import UsageService # noqa
+from ErrorReporter.error_report_presenter import ErrorReporterPresenter # noqa
+from ErrorReporter.errorreport import CrashReportPage # noqa
+
+
+def main():
+    if not UsageService.isEnabled():
```
Comment (martyngigg) [inline scripts/ErrorReporter/error_dialog_app.py]: There's a minor issue here that if you have usage reports disabled and the crash happens the exit code of the process is `0`. I think we need to make sure main always returns an integer and then main gets called as `sys.exit(main())` from below.

File: scripts/ErrorReporter/error_dialog_app.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        return
+    app = QtGui.QApplication(sys.argv)
+    form = CrashReportPage(show_continue_terminate=False)
+    ErrorReporterPresenter(form, command_line_args.exit_code)
+    app.exec_()
```
Comment (martyngigg) [inline scripts/ErrorReporter/error_dialog_app.py]: I think `return app.exec_()` is then what we want here also.

**Review by martyngigg — CHANGES_REQUESTED** [review]
This looks really good now. I think we just need to fix up the exit code situation from the app and then it's good to go.

**Review by martyngigg — APPROVED** [review]
This looks good - let's get it merged and out in the wild.

**Pull request merged**